rclpy/rclpy/parameter_client.py

```diff
@@ -0,0 +1,273 @@
+        :param callback:
+        :type callback: Union[Callable, None]
+        :return:
+        :rtype:
+        """
+        request = SetParameters.Request()
+        request.parameters = parameters
+        future = self.set_parameter_client_.call_async(request)
+        if callback:
+            future.add_done_callback(callback)
+        return future
+
+    def describe_parameters(
+        self, names: List[str], callback: Union[Callable, None] = None
+    ) -> Future:
+        """
+
+        :param names:
+        :type names: List[str]
+        :param callback:
+        :type callback: Union[Callable, None]
+        :return:
+        :rtype:
+
+        Parameter type definitions are given in rcl_interfaces.msg.ParameterType
+
+        """
+        request = DescribeParameters.Request()
+        request.names = names
+        future = self.describe_parameters_client_.call_async(request)
+        if callback:
+            future.add_done_callback(callback)
+        return future
+
+    def get_parameter_types(
+        self, names, callback: Union[Callable, None] = None
+    ) -> Future:
+        """
+
+        :param names:
+        :type names:
+        :param callback:
+        :type callback: Union[Callable, None]
+        :return:
+        :rtype:
+        """
+        request = GetParameterTypes.Request()
+        request.names = names
+        future = self.get_parameter_types_client_.call_async(request)
+        if callback:
+            future.add_done_callback(callback)
+        return future
+
+    def set_parameters_atomically(
+        self, parameters: Sequence[Parameter], callback: Union[Callable, None] = None
+    ) -> Future:
+        """
+
+        :param parameters:
+        :type parameters: Union[Set[Parameter], Parameter]
+        :param callback:
+        :type callback: Union[Callable, None]
+        :return:
+        :rtype:
+        """
+        request = SetParametersAtomically.Request()
+        request.parameters = parameters
+        future = self.set_parameters_atomically_client_.call_async(request)
+        if callback:
+            future.add_done_callback(callback)
+        return future
+
+    def delete_parameters(
+        self, names: List[str], callback: Union[Callable, None] = None
+    ) -> Future:
+        # TODO: `Static parameter cannot be undeclared`
+        """ Unset parameters with given names.
+
+        :param names:
+        :type names: List[str]
+        :param callback:
+        :type callback: Union[Callable, None]
+        :return:
+        :rtype:
+        """
+        request = SetParameters.Request()
+        request.parameters = [RclpyParameter(name=i).to_parameter_msg() for i in names]
+        future = self.set_parameter_client_.call_async(request)
+        if callback:
+            future.add_done_callback(callback)
+        return future
+
+    def load_parameter_file(self, parameter_file, use_wildcard, return_parameters=False):
```

Comment:
Similar to the c++ implementation, I think it would be nice to separate the logic of parsing of the parameters file and loading a dict of parameters.

Reply:
Loading a dict of parameters is just calling `set_parameters` on the parsed dict's values so I don't think there is too much of a point

Reply:
I think it might be more generally useful to have an API that reads an input parameters yaml and returns a dict. The implementation of such a function might be replaced by a call to an equivalent `rcl` function in the future.

Then it's true that this function's implementation becomes much simpler, but that's a good thing, IMO.

Reply:
For example, this function becomes something like:

```py
def load_parameter_file(self, parameter_file, use_wildcard, return_parameters=False):
    params_dict = load_parameters_yaml_file(parameter_file, use_wildcard)
    parameters = parse_parameter_dict(namespace='', parameter_dict=param_dict)
    return self.set_parameters(parameters)
```

Reply:
If done this way `params_dict` is largely useless since the parameters are still in string form at that point, `parse_parameter_dict` is what will turn them into a dict with ParameterMsg values. 

I agree that it would be nice to replace it with a `rcl` call. In the [rclcpp](https://github.com/ros2/rclcpp/blob/02802bcc385c3d6d814add93618f3388d44adec7/rclcpp/src/rclcpp/parameter_client.cpp#L295-L289) [implementaton](https://github.com/ros2/rclcpp/blob/02802bcc385c3d6d814add93618f3388d44adec7/rclcpp/src/rclcpp/parameter_map.cpp#L140-L130) it appears that the `rcl` call will directly return the parameter map which is what my [proposed approach](https://github.com/ros2/rclpy/blob/802adf1a36278fe3ce86a32e9c0a7dc47c786603/rclpy/rclpy/parameter.py#L265) does.